### DISCUSSION TITLE
Running new episodes until the first action is required?

### DISCUSSION THREAD

**Discussion by jacobtw2** [MarcEscandell/ALPypeRL#6] [General]
An issue I have noticed when trying to work with ALpypeRL is that the model needs to be in a position to perform the first action immediately upon start up. 

I have a model where I request an action each time a new order reaches a certain point in the flowchart logic and the action decides what to do with the order. Calling upon reset would cause an issue because there is no order that needs it yet

Is there a way we can tell the simulation to run until the first action before returning back to the python side when reset is called. i have tried calling runFast on startup but this did not work. 

Unfortunately, it is not even possible to manually set the model to a state where there is an order to process because any item for inserting an agent into a flow chart block such as inject or take is an event. Even though it happens in 0 time, it still requires time to move forward which it won't do until after going to the python side

Does anyone have any ideas on how to resolve this issue?

**Comment by MarcEscandell** [comment]
Hei Jacob!

I think I get your point, but let me rewrite it here in my own words just in case:

_ALPypeRL_ will call `takeAction` at `time() == 0` **always** to kickstart the model, and hence, `getObservation` and `getReward` as well. What you want is the simulation to first just run (no action executed yet) and wait until `requestAction()` is called (which may happen at `time() > 0` - later in the simulation). Please correct me if you don't mean exactly this.

If you do so, I think you are right. I have thought about this before. In the simulations/models I have build, the action was required at `time() == 0`, but I do understand this might not always be the case. I will add this _new feature_ requirement to the backlog. 

For now, the only recommendation I can give you to solve your issue is to add an _if-statement_ in your `takeAction`, `getObservation` and `getReward` functions. For example:
* **takeAction**:
```java
if (time() == 0) {
  // Do nothing
} else {
  // The rest of your code here
  // [...]
}
```
* **getObservation**:
```java
if (time() == 0) {
  // Just return an empty observation
  return <an-empty-observation>;
}
return <the-observation>;
```
* **getReward**:
```java
if (time() == 0) {
  // No reward
  return 0;
}
return <the-reward>;
```

The downside of this approach is that the neural network will associate a reward of 0 for any random action at the beginning. Maybe another good recommendation is to add `time()` as an entry in your _observation space_. That way the NN can match that behaviour to the simulation startup only.

As I said, I'll consider adding this feature in future releases. I'll keep you up to date. Happy coding!

**Reply by jacobtw2** [reply]
Hi Marc.

Thank you for the response. Yes that is what I meant. For now we have done something similar to your last suggestion and added an extra  parameter which is a 0/1 boolean flag for if this is the first setup action and then after that it will run as normal

**Comment by MarcEscandell** [comment]
Hei @jacobtw2! Please check the latest **alpyperl-1.0.0** version. I think I managed to fix this, but I need your input to validate it.

**Reply by jacobtw2** [reply]
Hi @MarcEscandell . I did some testing of the new version and added some tracelns. Could see that the reset function doesn't return until after the first time requestAction is called (5 seconds into the model) which is as expected. However, it still calls getObservation after resetting and does not call it again when the action is requested. 

This means that if the observation state changes between resetting and the first action being requested then it will not be correct

**Reply by MarcEscandell** [reply]
Hei @jacobtw2! You are absolutely right! 

`getObservation()` is still being called early at model launch (when objects in the AnyLogic model may have not been yet created, causing differences in size). 
I am not yet sure how to fix this problem. The reason why this method is called very early is because it is executed when `rllib` is creating the neural network and it needs to define the input layer size. For that, it queries AnyLogic during initialization.

For now, to fix this issue, I am just adding an if-statement and hardcoding the size. For example:
```java
// To avoid exception at model initialization
if (time() > 0) {
	List<Number> observation = new ArrayList<>();
	observation.addAll(car.getVisionReadings());
	observation.add(roundToDecimal(car.getLinearVelocity(), 2));
	observation.add(car.getTrackCompletion());
	observation.add(time());
	return observation.toArray(new Number[observation.size()]);
}
// Otherwise assume time=0 and return fix size
int size = (int) Math.sqrt(numSensors);
Integer[] emptyObs = new Integer[size * size + 3];
// Initialize all elements to zero
for (int i = 0; i < emptyObs.length; i++) {
    emptyObs[i] = 0;
}
return emptyObs;
```
I'll be thinking of a better way to fix this!

**Reply by MarcEscandell** [reply]
Just realized that I already wrote a similar idea some time ago. Sorry for repeating myself!
